src/data/data.cc

```diff
@@ -1051,6 +1063,13 @@ void SparsePage::SortIndices(int32_t n_threads) {
   });
 }
 
+void SparsePage::Reindex(uint64_t feature_offset, int32_t n_threads) {
+  auto& h_data = this->data.HostVector();
```

Comment:
This potentially pulls data from the device to the host.

Reply:
Agreed, but it's not much different from some of the other methods there.